atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java

```diff
@@ -250,4 +277,32 @@ public synchronized void shutdown() {
             throw Throwables.rewrapAndThrowUncheckedException(e);
         }
     }
+
+    private enum SweepOutcome {
+        SUCCESS, NOTHING_TO_SWEEP, DISABLED, UNABLE_TO_ACQUIRE_LOCKS,
+        NOT_ENOUGH_DB_NODES_ONLINE, TABLE_DROPPED_WHILE_SWEEPING, ERROR
+    }
+
+    private class SweepOutcomeMetrics {
+        private final MetricsManager metricsManager = new MetricsManager();
+        private final SlidingTimeWindowReservoir reservoir;
+
+        SweepOutcomeMetrics() {
+            Arrays.stream(SweepOutcome.values()).forEach(outcome ->
+                    metricsManager.registerMetric(BackgroundSweeperImpl.class, outcome.name(),
+                            () -> getOutcomeCount(outcome))
+            );
+            reservoir = new SlidingTimeWindowReservoir(60L, TimeUnit.SECONDS);
+        }
+
+        private Long getOutcomeCount(SweepOutcome outcome) {
+            return Arrays.stream(reservoir.getSnapshot().getValues())
+                    .filter(l -> l == outcome.ordinal())
```

Comment:
This is a bit strange, the metrics now report how many times an outcome was reported in the last minute, so SUCCESS will be 2 for some minute, 3 the next minute, 10 the next minute (if the batches are getting smaller/sweep is getting faster as backup just finished etc.), 

DISABLED will be reported much more frequently and the number could be 100?1000?

This could be useful info as we know how many times each status was reported in the last minute but this info is partially captured by SweepError metrics. Also, this makes setting up alerts a bit tricky as SUCCESS >=1 is success and NOT_ENOUGH_DB_NODES_ONLINE>=1 is we have at least one C* node unhealthy.

I don't directly see the benefit of reporting counts here and what kind of alerts we can have around them, but I maybe missing something. Should we make this 0-1 based on whether the metric was reported in the last minute?

Reply:
Also, for the disabled case, we can cause some memory overhead. See warning: http://metrics.dropwizard.io/3.1.0/manual/core/#man-core-histograms

Reply:
I don't think we need to be concerned about the memory overhead here.  We record a single integer each time sweep completes.  If sweep is disabled or the lock is out we pause for 2 minutes between runs - so we're actually only storing a single value.

If sweep succeeds we still sleep for 5 seconds so we're recording 12 metrics per second.  Maybe we'll reduce the sleep and speed up sweep - but we're a long way from recording 1000s of metrics between trimming the backing store.

Reply:
Having the counts allows us to see exactly what sweep has done.  Alarms can then be created for:
- success == 0 for all of the last 24 hours => no sweep run has succeeded, something is very bad
- db_unavailable > 0 for all of the last hour => sweep can't run because a db node is unhealthy and needs attention
- minimum across all hosts of lock_unavailable > 0 for more than 30 minutes => no node could get the sweep lock, implies it hasn't been properly cleared.

Since we can compare to 0, reporting just 0 or 1 doesn't give us anything over reporting the count and just loses information.